### DISCUSSION TITLE
Adding filter file

### DISCUSSION THREAD

**Discussion by getrude386** [mboquien/cigale#21] [General]
Hi, I am trying to add a filter which is not in the filters list (NB921), it's a Subaru narrow band filter. I get the following error, and below it is the file that I am trying to add:
![Screenshot 2022-07-30 at 23 52 23](https://user-images.githubusercontent.com/105718721/182327870-d948fd5b-25b8-4aa0-917f-b263459881fa.png)
![Screenshot 2022-07-30 at 23 53 12](https://user-images.githubusercontent.com/105718721/182327907-c190fd1a-aa19-47b0-88e1-d9f2a6a7016e.png)

**Comment by Guang91** [comment]
Are you using a Windows machine? Could you share the filter file? I can try on my Linux to see if it's an OS system.

**Reply by mboquien** [reply]
This is a fairly old version of CIGALE. Could you try with the latest version? The database is handled in a completely different manner now.

**Reply by getrude386** [reply]
Hi, I have now installed a new version, I'm still getting the same error.

Also, I am not using windows. I have attached the file.
[filter_NB921.txt](https://github.com/mboquien/cigale/files/9257779/filter_NB921.txt)

**Reply by Guang91** [reply]
Hello, I can successfully add that filter to my computer. 
Just to make sure, after you switched to v2022.1, you got the same error? 

![image](https://user-images.githubusercontent.com/7922328/182826626-d21d2b99-d762-4a33-b6bc-68f6f3336665.png)

**Reply by Guang91** [reply]
Another thing is the new version only works with Python 3.8 or higher.
Is your python version correct (your screenshot says it's 3.6)?

**Reply by getrude386** [reply]
Yes, I got the same error after switching to v2022.1. Let me update my python and try again.

**Reply by getrude386** [reply]
I am now using python3.8, and this happens:

![Screenshot 2022-08-05 at 09 39 16](https://user-images.githubusercontent.com/105718721/183027194-86beebc5-6748-4528-acfc-619904363230.png)

**Reply by Guang91** [reply]
It seems you don't have permission to write on the path listed at the end. 
Could you try to run as "sudo" and see?

**Reply by getrude386** [reply]
I am working inside a computing cloud called ilifu (slurm), so I just contacted the support team and asked them about it, this was their response:

You wont be able to add files to that path, if it's the path inside the container where cigale is installed. The containers are static and users shouldn't be able to modify them. If this is normal activity for the software it's likely that there is an alternative path or variable you'd able to set to point to your filter file. Maybe check cigale documentation, or you could download your own cigale package and use that instead of the one inside the container.

But I just told them that I am not using cigale from the container, I have downloaded it.

**Reply by Guang91** [reply]
There should be a way to install CIGALE to your local directory. @mboquien may know how to do it

**Reply by mboquien** [reply]
It depends on the details of the python installation. In my case `python setup.py develop --user` created a developer installation (that is you do not need to reinstall anything if you change CIGALE) with the executables located in `$HOME/.local/bin`. Once added to `$PATH`, then CIGALE runs from a local installation. You may want to remove any other CIGALE installation first though.

**Reply by getrude386** [reply]
The command (python setup.py develop --user) solved the problem. Thank you for your help.

**Reply by Guang91** [reply]
Great! I added a note to the installation page (https://github.com/mboquien/cigale/discussions/2) about the local installation.

**Reply by getrude386** [reply]
Is it normal that after adding the filter, when I check the filters list only 1 filter appears (the one I have added)?

**Reply by Guang91** [reply]
No, it's not normal. All default filters should be there. But I'm not sure about the case for local installation @mboquien

**Reply by getrude386** [reply]
> Great! I added a note to the installation page (#2) about the local installation.

Will be useful, thank you.

**Reply by mboquien** [reply]
It is not normal, indeed. I just tried adding a test filter and all the filters are there. When you do `pcigale-filters list` right after building the database, do you have a full list? The setup of your server seems problematic. It may be that there is a mixup somewhere and that CIGALE cannot find its database.

**Reply by getrude386** [reply]
No I don't get the full list, maybe I should just uninstall it and start again.

**Reply by mboquien** [reply]
It may be that there are conflicting CIGALE installations. You probably want to make sure to completely remove any trace of it (for instance `find ~ -name pcigale` should give you a number of locations where CIGALE is in your home directory.

**Reply by getrude386** [reply]
I have uninstalled and followed all the steps in #2 , But when I run command 7 I get this:

thando@ASTRO-PY3.8.simg:/idia/users/thando/cigale$ python setup.py build
/usr/local/lib/python3.8/dist-packages/setuptools/dist.py:545: UserWarning: The version specified ('dev') is an invalid version, this may not work as expected with newer versions of setuptools, pip, and PyPI. Please see PEP 440 for more details.
  warnings.warn(
running build
╭──────────────── Traceback (most recent call last) ────────────────╮
│ /idia/users/thando/cigale/setup.py:37 in <module>                 │
│                                                                   │
│   34 with open('pcigale/version.py') as f:                        │
│   35 │   exec(f.read())                                           │
│   36                                                              │
│ ❱ 37 setup(                                                       │
│   38 │   name="pcigale",                                          │
│   39 │   version=__version__,                                     │
│   40 │   packages=find_packages(exclude=["database_builder"]),    │
│                                                                   │
│ /usr/local/lib/python3.8/dist-packages/setuptools/__init__.py:87  │
│ in setup                                                          │
│                                                                   │
│    84 │   # Make sure we have any requirements needed to interpre │
│    85 │   logging.configure()                                     │
│    86 │   _install_setup_requires(attrs)                          │
│ ❱  87 │   return distutils.core.setup(**attrs)                    │
│    88                                                             │
│    89                                                             │
│    90 setup.__doc__ = distutils.core.setup.__doc__                │
│                                                                   │
.
.
.
.
.
.
.
ImportError: cannot import name 'SimpleDatabase' from 'pcigale.data' (unknown location)

**Reply by mboquien** [reply]
Can you provide the full backtrace?

**Reply by getrude386** [reply]
I now get so many different errors, I then decided to install the .tar file and install it, so now I get the following import error, tried to resolve it and failed:

Traceback (most recent call last):
  File "/usr/local/bin/pcigale-filters", line 5, in <module>
    from pcigale_filters import main
  File "/idia/users/thando/cigale-v2022.1/pcigale_filters/__init__.py", line 11, in <module>
    from pcigale.data import SimpleDatabase as Database
  File "/idia/users/thando/cigale-v2022.1/pcigale/__init__.py", line 18, in <module>
    from .session.configuration import Configuration
  File "/idia/users/thando/cigale-v2022.1/pcigale/session/configuration.py", line 12, in <module>
    from ..data import SimpleDatabase as Database
ImportError: cannot import name 'SimpleDatabase' from 'pcigale.data' (unknown location)

**Reply by mboquien** [reply]
I am at a loss as to what it may be. I do not remember seeing any error like this. I suspect there is some peculiarity the way the system is set up. I imagine this is a network drive? Could you try installing on a local drive?

Also, could you run this please?
```python
import sys

for path in sys.path:
    print(path)
```

**Reply by getrude386** [reply]
This is what I get from sys.path:

['/usr/local/bin',
 '/usr/lib/python38.zip',
 '/usr/lib/python3.8',
 '/usr/lib/python3.8/lib-dynload',
 '',
 '/users/thando/.local/lib/python3.8/site-packages',
 '/idia/users/thando/cigale-v2022.1',
 '/usr/local/lib/python3.8/dist-packages',
 '/opt/herschelhelp_python',
 '/opt/herschelhelp_internal',
 '/opt/XID_plus',
 '/usr/lib/python3/dist-packages']

**Reply by getrude386** [reply]
I don't know what went wrong because I had cigale and it was working, up until I updated to a new version and tried to add a new filter.

I have deleted pcigale from using this command: find ~ -name pcigale

But when I: "ls /usr/local/bin" I find pcigale

I have ran pcigale-filters list using the above path: "/usr/local/bin/pcigale-filters list", and I get the same error mention above (ImportError: cannot import name 'SimpleDatabase' from 'pcigale.data' (unknown location))

When I try to delete:
![Screenshot 2022-08-07 at 18 17 35](https://user-images.githubusercontent.com/105718721/183300626-7a172968-cabc-4757-b495-b85a0ed60db2.png)

**Reply by getrude386** [reply]
I just cleaned up the database again using number 6 in #2 and I now get the full list of filters, but when I try to add a new filter, I am getting the same error that I was getting before:

thando@ASTRO-PY3.8.simg:/idia/users/thando/cigale-v2022.1$ pcigale-filters add /idia/users/thando/filter_NB921.txt 
Importing NB921... (1001 points)
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /usr/local/bin/pcigale-filters:8 in <module>                                                     │
│                                                                                                  │
│   5 from pcigale_filters import main                                                             │
│   6 if __name__ == '__main__':                                                                   │
│   7 │   sys.argv[0] = re.sub(r'(-script\.pyw|\.exe)?$', '', sys.argv[0])                         │
│ ❱ 8 │   sys.exit(main())                                                                         │
│   9                                                                                              │
│                                                                                                  │
│ /idia/users/thando/cigale-v2022.1/pcigale_filters/__init__.py:154 in main                        │
│                                                                                                  │
│   151 │   │   if args.parser == 'list':                                                          │
│   152 │   │   │   list_filters()                                                                 │
│   153 │   │   elif args.parser == 'add':                                                         │
│ ❱ 154 │   │   │   add_filters(args.names)                                                        │
│   155 │   │   elif args.parser == 'plot':                                                        │
│   156 │   │   │   plot_filters(args.names)                                                       │
│   157                                                                                            │
│                                                                                                  │
│ /idia/users/thando/cigale-v2022.1/pcigale_filters/__init__.py:72 in add_filters                  │
│                                                                                                  │
│    69 │   │   else:                                                                              │
│    70 │   │   │   pivot = np.mean(wl[tr > 0.])                                                   │
│    71 │   │                                                                                      │
│ ❱  72 │   │   db.add({"name": name},                                                             │
│    73 │   │   │      {"wl": wl, "tr": tr, "pivot": pivot, "desc": desc})                         │
│    74 │   db.close()                                                                             │
│    75                                                                                            │
│                                                                                                  │
│ /usr/local/lib/python3.8/dist-packages/pcigale/data/__init__.py:123 in add                       │
│                                                                                                  │
│   120 │   │   entry = SimpleDatabaseEntry(primarykeys, data)                                     │
│   121 │   │   basename = "_".join(f"{k}={v}" for k, v in sorted(primarykeys.items()))            │
│   122 │   │                                                                                      │
│ ❱ 123 │   │   with open(self.path / Path(f"{basename}.pickle"), "wb") as f:                      │
│   124 │   │   │   pickle.dump(entry, f)                                                          │
│   125 │   │                                                                                      │
│   126 │   │   if len(self.parameters) == 0:  # Create the initial lists                          │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
OSError: [Errno 30] Read-only file system: '/usr/local/lib/python3.8/dist-packages/pcigale/data/filters/name=NB921.pickle'



When I ran number 7 in #2 (python setup.py build):

Traceback (most recent call last):
  File "setup.py", line 34, in <module>
    with open('pcigale/version.py') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'pcigale/version.py'

**Reply by mboquien** [reply]
It really looks like you have conflicting installations of CIGALE. Compare the path of  `/idia/users/thando/cigale-v2022.1/pcigale_filters/__init__.py` versus `/usr/local/lib/python3.8/dist-packages/pcigale/data/__init__.py` in the backtrace. This should not happen. I would suggest to delete all the pcigale-related packages in `/usr/local/lib/python3.8/dist-packages` and then redo an installation.

**Reply by getrude386** [reply]
thando@ASTRO-PY3.8.simg:/idia/users/thando$ rm -r /usr/local/lib/python3.8/dist-packages/pcigale*
rm: cannot remove '/usr/local/lib/python3.8/dist-packages/pcigale': Read-only file system
rm: cannot remove '/usr/local/lib/python3.8/dist-packages/pcigale-2022.0.dist-info': Read-only file system
rm: cannot remove '/usr/local/lib/python3.8/dist-packages/pcigale_filters': Read-only file system
rm: cannot remove '/usr/local/lib/python3.8/dist-packages/pcigale_plots': Read-only file system

**Reply by getrude386** [reply]
I think maybe the ones in this path /usr/local/lib/python3.8/dist-packages are contained in ilifu container,

**Reply by mboquien** [reply]
Yes, this may be causing a conflict. Best would be to try removing this container first.

**Reply by getrude386** [reply]
I want to try to reinstall it outside the container.

**Reply by getrude386** [reply]
I have ran all the commands in #2 , they worked successfully. When I run "pcigale genconf" I get:

                           ╭───────────────────────────────────────────────────────────────────────╮                            
                           │                  Code Investigating GALaxy Emission                   │                            
                           │             Boquien et al. (2019) (https://cigale.lam.fr)             │                            
                           │ CIGALE version: dev — Python version: 3.8.10 — Platform: linux-x86_64 │                            
                           ╰───────────────────────────────────────────────────────────────────────╯                            
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /users/thando/.local/bin/pcigale:11 in <module>                                                  │
│                                                                                                  │
│    8 if __name__ == '__main__':                                                                  │
│    9 │   sys.argv[0] = re.sub(r'(-script\.pyw?|\.exe)?$', '', sys.argv[0])                       │
│   10 │   sys.exit(                                                                               │
│ ❱ 11 │   │   load_entry_point('pcigale', 'console_scripts', 'pcigale')()                         │
│   12 │   )                                                                                       │
│   13                                                                                             │
│                                                                                                  │
│ /idia/users/thando/cigale/pcigale/__init__.py:139 in main                                        │
│                                                                                                  │
│   136 │   │   if args.parser == "init":                                                          │
│   137 │   │   │   init(config)                                                                   │
│   138 │   │   elif args.parser == "genconf":                                                     │
│ ❱ 139 │   │   │   genconf(config)                                                                │
│   140 │   │   elif args.parser == "check":                                                       │
│   141 │   │   │   check(config)                                                                  │
│   142 │   │   elif args.parser == "run":                                                         │
│                                                                                                  │
│ /idia/users/thando/cigale/pcigale/__init__.py:41 in genconf                                      │
│                                                                                                  │
│    38                                                                                            │
│    39 def genconf(config):                                                                       │
│    40 │   """Generate the full configuration."""                                                 │
│ ❱  41 │   config.generate_conf()                                                                 │
│    42 │   console.print(                                                                         │
│    43 │   │   f"{INFO} The configuration file has been updated. Please complete the "            │
│    44 │   │   "various module parameters and the data file columns to use in the "               │
│                                                                                                  │
│ /idia/users/thando/cigale/pcigale/session/configuration.py:308 in generate_conf                  │
│                                                                                                  │
│   305 │   │   self.spec["analysis_params"] = {}                                                  │
│   306 │   │                                                                                      │
│   307 │   │   module_name = self.config["analysis_method"]                                       │
│ ❱ 308 │   │   for name, (typ, desc, default) in analysis_modules.get_module(                     │
│   309 │   │   │   module_name, blank=True                                                        │
│   310 │   │   ).parameters.items():                                                              │
│   311 │   │   │   if default is None:                                                            │
│                                                                                                  │
│ /idia/users/thando/cigale/pcigale/analysis_modules/__init__.py:130 in get_module                 │
│                                                                                                  │
│   127 │                                                                                          │
│   128 │   try:                                                                                   │
│   129 │   │   module = import_module('.' + module_name, 'pcigale.analysis_modules')              │
│ ❱ 130 │   │   return module.Module(**kwargs)                                                     │
│   131 │   except ImportError:                                                                    │
│   132 │   │   raise Exception(f"Module {module_name} could not be imported.")                    │
│   133                                                                                            │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: module 'pcigale.analysis_modules' has no attribute 'Module'

**Reply by getrude386** [reply]
But, great news is that I have now managed to add my filter and I can also see the full list of filters after adding the new one.

**Reply by getrude386** [reply]
Hi, It seems like I did not do the correct thing, I get this error:

pcigale.data.DatabaseLookupError: The filter <NB921> is not in the database


Whereas when I check the filters list I do get it:

thando@slurm-login:/scratch3/users/thando/cigale-v2022.1$ pcigale-filters list | grep -i NB9
                 NB921                                                         wavelength[angstrom]  transmittance              920   1001

**Reply by mboquien** [reply]
It seems it was not imported correctly. Look at the filter format in `database_builder/filters` for examples.

**Reply by getrude386** [reply]
I still get the same error, I did not spot any differences when comparing to the examples.

**Reply by mboquien** [reply]
I did not realize that your filter description was `wavelength[angstrom] transmittance]`. So it looks good. However I am puzzled about your exception. It should not happen as this corresponds to the database library of an old version of CIGALE. So my guess is that you still have a mix of CIGALE versions.

**Reply by getrude386** [reply]
Ok, I will try to install it again from the scratch and see.

**Comment by wilnze** [comment]
@Guang91, all data files with .fit always come out with no SED error.
I am using the latest Cigale and I cannot find a solution to it

**Reply by Guang91** [reply]
Do you mean the best-fit SED files? Those are only the model with minimum chi^2, without well-defined errors. You can only have errors for physical quantities and band fluxes.
